src/main/java/org/jcodings/transcode/TranscodingManager.java

```diff
@@ -237,37 +235,42 @@ private static EConv open0(byte[] sourceEncodingName, byte[] destinationEncoding
 
     @SuppressWarnings("unchecked")
     @TruffleBoundary
-    private static LinkedList<String> bfs(String sourceEncodingName, String destinationEncodingName) {
-        final Set<String> alreadyVisited = new HashSet<>();
-        final Deque<LinkedList<String>> queue = new ArrayDeque<>();
-
-        LinkedList<String> path = new LinkedList<>();
-        path.add(sourceEncodingName);
-        queue.add(path);
-        alreadyVisited.add(sourceEncodingName);
-
-        while (! queue.isEmpty()) {
-            path = queue.pop();
-            final String sourceName = path.getLast();
-
-            if (allTranscoders.get(sourceName).containsKey(destinationEncodingName)) {
-                path.add(destinationEncodingName);
-                return path;
-            } else {
-                for (String destinationName : allTranscoders.get(sourceName).keySet()) {
-                    if (alreadyVisited.contains(destinationName)) {
-                        continue;
+    public static LinkedList<String> bfs(String sourceEncodingName, String destinationEncodingName) {
+        final Deque<String> queue = new ArrayDeque<>();
+        final HashMap<String, String> invertedList = new HashMap<>();
+
+        invertedList.put(sourceEncodingName, null);
+
+        queue.add(sourceEncodingName);
+        while (!queue.isEmpty()) {
+            String current = queue.pop();
+
+            for (String child : allTranscoders.get(current).keySet()) {
+                if (invertedList.containsKey(child)) {
+                    // We've already visited this path or are scheduled to.
+                    continue;
+                }
+
+                if (child.equals(destinationEncodingName)) {
+                    // Search finished.
+                    final LinkedList<String> ret = new LinkedList<>();
+                    ret.add(child);
+
+                    String next = current;
+                    while (next != null) {
+                        ret.addFirst(next);
+                        next = invertedList.get(next);
                     }
 
-                    final LinkedList<String> newPath = (LinkedList<String>) path.clone();
-                    newPath.add(destinationName);
-                    queue.add(newPath);
-                    alreadyVisited.add(destinationName);
+                    return ret;
                 }
+
+                invertedList.put(child, current);
+                queue.add(child);
             }
         }
 
-        return null;
+        return new LinkedList<>();
```

Comment:
We could relax it. There's one place that expects `remove` to be a fast operation, but that could be rewritten. In practice, we don't often hit this branch anyway.